src/modules/Game.class.js

```diff
@@ -1,68 +1,263 @@
+    initialState = Array.from({ length: 4 }, () => Array(4).fill(0)),
+  ) {
+    this.initialState = initialState;
+    this.state = initialState.map((row) => [...row]);
+    this.status = Game.gameStatuses.idle;
+    this.score = 0;
+  }
+
+  moveLeft() {
+    if (!this.isStateValid(this.state)) {
+      return;
+    }
+
+    const updatedState = this.state.map((row) => this.applyMove(row));
+
+    this.updateGameState(updatedState);
+    this.completeMoveTasks();
+  }
+
+  moveRight() {
+    const reversedState = this.state.map((row) => [...row].reverse());
+
+    if (!this.isStateValid(reversedState)) {
+      return;
+    }
+
+    const updatedState = reversedState.map((row) =>
+      // eslint-disable-next-line prettier/prettier
+      this.applyMove(row).reverse());
+
+    this.updateGameState(updatedState);
+    this.completeMoveTasks();
+  }
+
+  moveUp() {
+    const rotatedRightState = this.rotateRight(this.state);
+
+    if (!this.isStateValid(rotatedRightState)) {
+      return;
+    }
+
+    const newState = [
+      ...rotatedRightState.map((row) => this.applyMove([...row])),
+    ];
+
+    const rotatedLeftState = this.rotateLeft(newState);
+
+    this.updateGameState(rotatedLeftState);
+    this.completeMoveTasks();
+  }
+
+  moveDown() {
+    const rotatedRightState = this.rotateRight(this.state);
+    const rotatedLocaleState = [
+      ...rotatedRightState.map((row) => [...row].reverse()),
+    ];
+
+    if (!this.isStateValid(rotatedLocaleState)) {
+      return;
+    }
+
+    const newState = rotatedLocaleState.map((row) =>
+      // eslint-disable-next-line prettier/prettier
+      this.applyMove([...row]).reverse());
+    const rotatedLeftState = this.rotateLeft(newState);
+
+    this.updateGameState(rotatedLeftState);
+    this.completeMoveTasks();
+  }
+
+  applyMove(vector) {
+    const newRow = [];
+
+    let i = 0;
+
+    while (i < vector.length) {
+      const current = vector[i];
+      const next = vector[i + 1];
+
+      if (current) {
+        if (current === next) {
+          newRow.push(current * 2);
+          this.score += current * 2;
+          i += 2;
+        } else {
+          newRow.push(current);
+          i++;
+        }
+      } else {
+        i++;
+      }
+    }
+
+    while (newRow.length < vector.length) {
+      newRow.push(0);
+    }
+
+    return newRow;
+  }
+
+  getScore() {
+    return this.score;
+  }
+
+  getState() {
+    return this.state;
+  }
+
+  getStatus() {
+    return this.status;
+  }
+
+  start() {
+    this.status = Game.gameStatuses.playing;
+    this.completeMoveTasks(2);
+  }
+
+  restart() {
+    this.resetState();
+    this.status = Game.gameStatuses.idle;
+    this.score = 0;
+  }
+
+  generateNewTile() {
+    const emptyCells = this.getEmptyCells();
+
+    if (!emptyCells.length) {
+      return;
+    }
+
+    const [row, col] =
+      emptyCells[Math.floor(Math.random() * emptyCells.length)];
+
+    this.state[row][col] = Math.random() < 0.9 ? 2 : 4;
```

Comment:
The random tile generation is not deterministic, which makes the game logic difficult to test. Consider passing in a random number generator or a function that can be mocked or seeded for consistent testing purposes.